# Assignment 2 (CSI 4142) [Part 1]

# Group Information

- **Student name**: Shabrina Sharmin
- **Student Id**: 300230297
- **Group no**: A-85

# Clean Data checking

## Description of the dataset1 (Cafe Sales - Dirty Data for Cleaning Training)

The dataset contains various information about cafe sales. The dataset has many missing, and inconsistent information as well. This dataset is a perfect fit for cleaning techniques, data wrangling, and feature engineering.
A complete description of the features the dataset contains is provided below:
* Transaction ID **(object)**: A unique identifier for each transaction.
* Item **(object)**: The name of the item purchased.
* Quantity **(object)**: The quantity of the item purchased.
* Price Per Unit **(object)**: The price of a single unit of the item.
* Total Spent **(object)**: The total amount spent on the transaction.
* Payment Method  **(object)**: The method of payment used.
* Location **(object)**: The location where the transaction occurred.
* Transaction Date **(object)**: The date of the transaction.



## Description of dataset2 (Movies Dataset for Feature Extraction, Prediction)

The dataset contains certain information regarding diffrent movies and tvshows which are top in netflix. A complete description of the features of the dataset is provided below:
* MOVIES **(object)**: The movie names are given under the column
* YEAR **(object)**: The year of movie or tv shows telecast for audience
* GENRE **(object)**: The dataset contain many genres most valuable for recommendation system
* RATING **(float64)**: The audience thought about movie or tv show given
* ONE-LINE **(object)**: The short description about movie or tv show for audiences first impression
* STARS **(object)**: The casting of art is define under this columns which makes talk between audience
* VOTES **(object)**: The audience express their view under the columns its useful to identify impact make by art
* RunTime **(float64)**: The duration for runing time of art
* Gross **(object)**: It gives the total amount earned in worldwide

## **Access data**

### **Local instruction for accessing dataset on kaggle:** ###
**Pre-Requisite:** User must respectively create and download their access token from https://www.kaggle.com/. If the user already has an existing kaggle token, then no need to create a new one. Just use the existing kaggle token (kaggle.json) as requested.
If the "kaggle.json" token is not under ".kaggle" folder in the home path, then place the kaggle.json in the same directoy as the notebook. The code below will copy the "kaggle.json" token to the correct location

In [1]:
# Create a kaggle configuration folder/directory
!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/
!chmod 600 ~/.kaggle/kaggle.json  # Adjust the permissions of the file.

cp: cannot stat 'kaggle.json': No such file or directory
chmod: cannot access '/root/.kaggle/kaggle.json': No such file or directory


### **Instruction for running in Google Colab** ###

Upon running the code in following section below, a prompt will pop up asking for the **'kaggle.json'** file. The user needs to click on the 'Choose file' option and navigate to the corresponding directory where the kaggle.json file should reside and choose **'kaggle.json'** file. At this point, the "kaggle.json" should be loaded to the instance of the colab.

In [2]:
from os import mkdir

## Upload kaggle API credentials to Colab
from google.colab import files
files.upload()

Saving kaggle.json to kaggle.json


{'kaggle.json': b'{"username":"shabrinasharmin","key":"62fe625309926cb95271b826f7908041"}'}

### **Setting up the appropriate libraries needed for the assignment** ##
Run the follwing code below to set up the appropriate libraries that are needed

In [3]:
## Install Kaggle library
!pip install -q kaggle
!pip install pandas
!pip install seaborn
!pip install matplotlib
!pip install numpy

### **Download the dataset**

In [4]:
## Download the dataset 1 (dirty_cafe_sales.csv)
!kaggle datasets download -d ahmedmohamed2003/cafe-sales-dirty-data-for-cleaning-training
## Unzip the dataset 1
!unzip cafe-sales-dirty-data-for-cleaning-training.zip

## download dataset 2 (movies.csv)
!kaggle datasets download -d bharatnatrayn/movies-dataset-for-feature-extracion-prediction
## unzip the dataset 2
!unzip movies-dataset-for-feature-extracion-prediction.zip

Dataset URL: https://www.kaggle.com/datasets/ahmedmohamed2003/cafe-sales-dirty-data-for-cleaning-training
License(s): CC-BY-SA-4.0
  0% 0.00/111k [00:00<?, ?B/s]
100% 111k/111k [00:00<00:00, 56.8MB/s]
Archive:  cafe-sales-dirty-data-for-cleaning-training.zip
  inflating: dirty_cafe_sales.csv    
Dataset URL: https://www.kaggle.com/datasets/bharatnatrayn/movies-dataset-for-feature-extracion-prediction
License(s): CC0-1.0
  0% 0.00/1.05M [00:00<?, ?B/s]
100% 1.05M/1.05M [00:00<00:00, 141MB/s]
Archive:  movies-dataset-for-feature-extracion-prediction.zip
  inflating: movies.csv              


## Imports

In [5]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import datetime
from datetime import datetime
from datetime import date

## Read the files

In [6]:
# read the first dataset (dirty_cafe_sales>csv)
data_dirty_cafe_sales = pd.read_csv("dirty_cafe_sales.csv")
data_dirty_cafe_sales.head()
data_dirty_cafe_sales.info()
data_dirty_cafe_sales.describe()
data_dirty_cafe_sales.shape
data_dirty_cafe_sales.notnull

#Read the second dateset (movies.csv)
data_movies = pd.read_csv("movies.csv")
# data_movies.head()
data_movies.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction ID    10000 non-null  object
 1   Item              9667 non-null   object
 2   Quantity          9862 non-null   object
 3   Price Per Unit    9821 non-null   object
 4   Total Spent       9827 non-null   object
 5   Payment Method    7421 non-null   object
 6   Location          6735 non-null   object
 7   Transaction Date  9841 non-null   object
dtypes: object(8)
memory usage: 625.1+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9999 entries, 0 to 9998
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   MOVIES    9999 non-null   object 
 1   YEAR      9355 non-null   object 
 2   GENRE     9919 non-null   object 
 3   RATING    8179 non-null   float64
 4   ONE-LINE  9999 non-null   object 
 5   STARS  

In [ ]:
# from google.colab import drive
# drive.mount('/content/drive')

## Data Clean Checker

## Type of errors

## Error no:1 (Data type checking)

### **Cell 1: Test Name and description**
  * Error name: **Data Type Error**
  * Error description: This test ensures that each column has the correct type of data. Data type miss-match can lead to errors in data manipulation and thus can results in an ineffective analytical result.

### **Cell 2: Parameters for checker**

In [8]:
## parameter for data type checker
expected_types = {
    'Transaction ID': 'object',  # Expected to be string, which is 'object' in pandas
    'Item': 'object',            # Expected to be string
    'Quantity': 'Int64',         # Expected to be integer
    'Price Per Unit': 'float64', # Expected to be float
    'Total Spent': 'float64',    # Expected to be float
    'Payment Method': 'object',  # Expected to be string
    'Location': 'object',        # Expected to be string
    'Transaction Date': 'datetime64[ns]'  # Expected to be datetime
}

### **Cell 3: Checker code**

In [9]:
## function to check type
def check_data_types(df, expected_types):
  error_type ={}
  for col, expected_type in expected_types.items():
    actual_type = df[col].dtype
    if str(actual_type) != expected_type:
      error_type[col] = f"Expected: {expected_type}, Found: {actual_type}"

  return error_type


## func to display type error
def display_type_error(data_type):
  if data_type:
    for column, error in data_type.items():
        print(f"Data type mismatch in column '{column}': {error}")
  else:
    print("All column data types are correct.")

data_type_check = check_data_types(data_dirty_cafe_sales, expected_types)
display_type_error(data_type_check)

Data type mismatch in column 'Quantity': Expected: Int64, Found: object
Data type mismatch in column 'Price Per Unit': Expected: float64, Found: object
Data type mismatch in column 'Total Spent': Expected: float64, Found: object
Data type mismatch in column 'Transaction Date': Expected: datetime64[ns], Found: object


### **Cell 4: Report of findings**

For the dataset, the columns **' Quantity '**  **' Price Per Unit '**, and **' Total Spent '** should be in numerical form; whereas we found them to be of object type (please, refer to the output of the previous cell). Therefore, indicating the necessity of pre processing some columns for effective and meaningful analysis of the dataset.

## Error no:2 (Range checking)

### **Cell 1: Test Name and description**
  * Error name: **Range checking Error**
  * Error description: This test ensures that each column has value that corresponds to a specific attribute is within the range. This error checking will display if the specified attribute has any out of bound. Here, any column that has Unknown/Error/Empty value has been considered as out of bound.

### **Cell 2: Parameters for checker**

In [10]:
# #data pre-process
df = data_dirty_cafe_sales.copy()

#convert col 'price per unit' to numeric(float)
df['Price Per Unit']  = pd.to_numeric(df['Price Per Unit'], errors='coerce')

expected_menu_prices = {
    "Coffee": 2,
    "Tea": 1.5,
    "Sandwich": 4,
    "Salad": 5,
    "Cake": 3,
    "Cookie": 1,
    "Smoothie": 4,
    "Juice": 3
}
 #Item for which want to check the range
item_to_Check = 'Tea'

### **Cell 3: Checker code**

In [12]:
#func to check if prices for menu items are in the correct range
def check_menu_prices(row, expected_menu, item):
  item_in_row = row['Item']
  expected_item_price = expected_menu.get(item)
  actual_price_per_unit = row['Price Per Unit']
  #check if actual and expected prices are same
  if item_in_row == item:
    if expected_item_price != actual_price_per_unit or actual_price_per_unit <= 0:
      return f"Expected: {expected_item_price}, Found: {actual_price_per_unit}"
  return None

 #apply the func check_menu_prices
df['Price Check'] = df.apply(lambda row: check_menu_prices(row, expected_menu_prices, item_to_Check), axis=1)

# Filter out only mismatched prices
mismatched_prices = df[df['Price Check'].notna()]

# Display findings
print("Mismatched Prices:")
print(mismatched_prices[['Item', 'Price Per Unit', 'Price Check']])
if len(mismatched_prices) > 0:
  print(f"There are {len(mismatched_prices)} mismatched prices for the Item {item_to_Check}")
else:
  print("No mismatched prices found.")



Mismatched Prices:
     Item  Price Per Unit                Price Check
85    Tea             NaN  Expected: 1.5, Found: nan
196   Tea             NaN  Expected: 1.5, Found: nan
333   Tea             NaN  Expected: 1.5, Found: nan
441   Tea             NaN  Expected: 1.5, Found: nan
757   Tea             NaN  Expected: 1.5, Found: nan
...   ...             ...                        ...
8970  Tea             NaN  Expected: 1.5, Found: nan
8975  Tea             NaN  Expected: 1.5, Found: nan
9212  Tea             NaN  Expected: 1.5, Found: nan
9295  Tea             NaN  Expected: 1.5, Found: nan
9961  Tea             NaN  Expected: 1.5, Found: nan

[66 rows x 3 columns]
There are 66 mismatched prices for the Item Tea


### **Cell 4: Report of findings**
From the output of our range checking, we have identified mismatched entries for the "Item" **Tea**. Here, "mismatched" refers to prices that are either less than, greater than the expected price of $1.5, or where the price data is missing altogether. The checker also identifies any negative values, as such prices are not feasible and indicate data entry errors.

The output provides a snapshot of 66 rows showing discrepancies, predominantly with "NaN" as the unit price for Tea. This highlights a significant issue with data entries, suggesting that many records are incomplete or improperly recorded. These findings underscore the necessity for a thorough data cleaning process to prepare the dataset for effective analysis.

This data cleaning will involve confirming the source of these NaN values and correcting them where possible, or deciding on an appropriate handling strategy such as imputation or exclusion, depending on the analysis needs and data integrity standards.


## Error no:3 (Format errors)

### **Cell 1: Test Name and description**
  * Error name: **Format Error**
  * Error description: This error occurs when entries in the 'Transaction Date' column do not adhere to the specified date format ('%Y-%m-%d'). The dataset contains a mixture of entries marked as 'ERROR' and 'UNKNOWN', which indicate formatting inconsistencies and missing or null values, respectively

### **Cell 2: Parameters for checker**


In [13]:
#format for date
date_format = '%Y-%m-%d'

### **Cell 3: Checker code**

In [14]:
# #func to check date format
def check_date_format(row, date_format):
  date_to_check = row['Transaction Date']
  try:
    #try to convert the date_to_check according to the expected format
    valid_date = pd.to_datetime(date_to_check, format=date_format, exact=True)
    return None
  except ValueError:
    return f"Invalid date format: {date_to_check}"

## Apply the func check_date_format ro each row in the dataframe
df['Date Format Check'] = df.apply(lambda row: check_date_format(row, date_format), axis=1)

#filter out incorrectly formatted date
incorrect_format_dates = df[df['Date Format Check'].notna()]

# display findings
print("Incorrectly Formatted Dates:")
print(incorrect_format_dates[['Transaction ID','Transaction Date', 'Date Format Check']])
if len(incorrect_format_dates) > 0:
  print(f"There are {len(incorrect_format_dates)} incorrectly formatted dates.")
else:
  print("No incorrectly formatted dates found.")


Incorrectly Formatted Dates:
     Transaction ID Transaction Date             Date Format Check
11      TXN_3051279            ERROR    Invalid date format: ERROR
29      TXN_7640952            ERROR    Invalid date format: ERROR
33      TXN_7710508            ERROR    Invalid date format: ERROR
103     TXN_7028009            ERROR    Invalid date format: ERROR
115     TXN_1001832          UNKNOWN  Invalid date format: UNKNOWN
...             ...              ...                           ...
9907    TXN_8104914          UNKNOWN  Invalid date format: UNKNOWN
9933    TXN_9460419          UNKNOWN  Invalid date format: UNKNOWN
9937    TXN_8253472          UNKNOWN  Invalid date format: UNKNOWN
9949    TXN_3130865          UNKNOWN  Invalid date format: UNKNOWN
9983    TXN_9226047          UNKNOWN  Invalid date format: UNKNOWN

[301 rows x 3 columns]
There are 301 incorrectly formatted dates.


### **Cell 4: Report of findings**

From the output displayed above we can see the list of transaction dates that are not in the correct/specified format in the dataset. There are about 301 rows in the dataset where the value for column **'Transaction Date'** is not in the expected format. As we can see in the generated output that, there are values with 'UNKNOWN', and 'ERROR'; which can not be interpreted as the expected date format. This indicates the necessity of cleanig dataset with an effective approch so that an meaningful analysis can be done.

## Error no:4 (Consistency errors)


### **Cell 1: Test Name and description**
  * Error name: **Consistency Error**
  * Error description: This error identifies discrepancies between the recorded "Total Spent" and the product of "Price Per Unit" and "Quantity" for each transaction in the dataset.If the total computed from the unit price and quantity does not match the recorded total spent, it indicates a potential issue that could impact financial reporting, inventory tracking, and business analytics along with inconsistency in the dataset.

### **Cell 2: Parameters for checker**

In [16]:

# pre process the specific cols
df['Quantity'] = pd.to_numeric(df['Quantity'], errors='coerce')
df['Quantity'] = df['Quantity'].astype('Int64')   # quantity can not be a fraction
df['Total Spent']  = pd.to_numeric(df['Total Spent'], errors='coerce')
# df.info()
# menu items prices
expected_menu_prices = {
    "Coffee": 2,
    "Tea": 1.5,
    "Sandwich": 4,
    "Salad": 5,
    "Cake": 3,
    "Cookie": 1,
    "Smoothie": 4,
    "Juice": 3
}
 #Item for which want to check the range
item_to_Check_consistency = 'Tea'

### **Cell 3: Checker code**

In [17]:
#func to check consistency
def check_consistency(row, expected_menu, item):
  item_in_row = row['Item']
  # check if the row contains the expected item
  if item_in_row == item:
    expected_item_price = expected_menu.get(item)
    actual_price_per_unit = row['Price Per Unit']
    actual_total_spent = row['Total Spent']

    # check if all data are present and in valid format
    if(pd.notna(row['Quantity']) and pd.notna(row['Price Per Unit'])):
      # check if Quantity and Price Per Unit is positive
      if (row['Quantity'] > 0  and  row['Price Per Unit'] > 0):
        expected_total_spent = row['Quantity'] * row['Price Per Unit']
        # check id actual and exopected total amount matches
        if expected_total_spent != actual_total_spent:
          return f"Expected: {expected_total_spent}, Found: {actual_total_spent}"
      else:
        return f"Quantity or Price Per Unit is either zero or negative"
    else:
      return f"Quantity or Price Per Unit is missing in dataframe"
  else:
    return None

#apply the fuc to dataset for consistency checking
df['Consistency Check'] = df.apply(lambda row: check_consistency(row, expected_menu_prices, item_to_Check_consistency), axis=1)

#filter out the inconsistenyt record
inconsistencies = df[df['Consistency Check'].notna()]

# Display findings
print("Inconsistencies:")
print(inconsistencies[['Transaction ID','Item', 'Quantity', 'Price Per Unit', 'Total Spent', 'Consistency Check']])
if len(inconsistencies) > 0:
  print(f"There are {len(inconsistencies)} inconsistencies for the Item {item_to_Check_consistency}")
else:
  print("No inconsistencies found.")


Inconsistencies:
     Transaction ID Item  Quantity  Price Per Unit  Total Spent  \
42      TXN_6650263  Tea         2             1.5          NaN   
85      TXN_8035512  Tea         3             NaN          4.5   
171     TXN_8989148  Tea         2             1.5          NaN   
196     TXN_3800182  Tea         2             NaN          3.0   
253     TXN_3544789  Tea         4             1.5          NaN   
...             ...  ...       ...             ...          ...   
9797    TXN_4518505  Tea      <NA>             1.5          7.5   
9832    TXN_3325234  Tea      <NA>             1.5          1.5   
9932    TXN_8502079  Tea      <NA>             1.5          3.0   
9935    TXN_9778251  Tea      <NA>             1.5          6.0   
9961    TXN_2153100  Tea         2             NaN          3.0   

                                      Consistency Check  
42                            Expected: 3.0, Found: nan  
85    Quantity or Price Per Unit is missing in dataf...  
171 

### **Cell 4: Report of findings**
Based on the generated output above, the consistency check has identified 161 transactions for the item "Tea" with discrepancies between the expected and recorded 'Total Spent', primarily due to mismatched calculations where both 'Quantity' and 'Price Per Unit' are present, and due to missing data in either of these fields, which prevents accurate calculation. These inconsistencies, which include several cases where calculated totals differ from recorded totals and instances of missing data (e.g., non-null 'Total Spent' with missing 'Quantity' or 'Price Per Unit'), pose risks to financial reporting and inventory management. In order to conduct an accurate data analysis and gain an accurate insight, a systematic data cleaning process is necessary here.

## Error no:5 (Uniqueness errors)

### **Cell 1: Test Name and description**
  * Error name: **Uniqueness Error**
  * Error description: This test ensures that the fields for each row in a column is unique; meaning there is no duplicte values under a column. For this test, we used 'Transaction Date' column.

### **Cell 2: Parameters for checker**

In [18]:
unique_params = ['Transaction ID' 'Transaction Date']
param_to_check_unique = 'Transaction Date'

### **Cell 3:  Checker code**

In [21]:
# func to check uniqueness
def check_uniqueness(param_to_check):
  if(param_to_check == 'Transaction Date'):
    df['Transaction Date'] = pd.to_datetime(df['Transaction Date'], errors='coerce')
  ids =  df[param_to_check].unique()
  print("There are",len(ids),"unique", param_to_check ,"in the dataset")
  print("The first 100", param_to_check, "are printed below for visualization purpose:")
  print(ids[:100])

#call the func check_uniqueness
check_uniqueness(param_to_check_unique)


There are 366 unique Transaction Date in the dataset
The first 100 Transaction Date are printed below for visualization purpose:
<DatetimeArray>
['2023-09-08 00:00:00', '2023-05-16 00:00:00', '2023-07-19 00:00:00',
 '2023-04-27 00:00:00', '2023-06-11 00:00:00', '2023-03-31 00:00:00',
 '2023-10-06 00:00:00', '2023-10-28 00:00:00', '2023-07-28 00:00:00',
 '2023-12-31 00:00:00', '2023-11-07 00:00:00',                 'NaT',
 '2023-05-03 00:00:00', '2023-06-01 00:00:00', '2023-03-21 00:00:00',
 '2023-11-15 00:00:00', '2023-06-10 00:00:00', '2023-02-24 00:00:00',
 '2023-03-25 00:00:00', '2023-01-15 00:00:00', '2023-04-04 00:00:00',
 '2023-03-30 00:00:00', '2023-12-01 00:00:00', '2023-09-18 00:00:00',
 '2023-06-03 00:00:00', '2023-12-13 00:00:00', '2023-04-20 00:00:00',
 '2023-04-10 00:00:00', '2023-03-11 00:00:00', '2023-06-02 00:00:00',
 '2023-11-06 00:00:00', '2023-08-15 00:00:00', '2023-10-09 00:00:00',
 '2023-05-28 00:00:00', '2023-07-17 00:00:00', '2023-04-29 00:00:00',
 '2023-06-08 00

### **Cell 4: Report of findings**
From the output of the uniqueness checker code above, we can see that, there are 366 unique dates under the column 'Transaction Date'. The dataset has 10,000 rows and thus there are numerous transaction date that are not unique. However, this makes sense for this dataset. Since, the dataset contain sales information of a coffee shop and multiple sales are supposed to take place on same date. Thus, from the domain knowledge, we understand the reason of the column 'Transaction Date' not being unique.

## Error no:6 (Presence errors)

### **Cell 1: Test Name and description**
  * Error name: **Presence Error**
  * Error description: A "presence error" checker for the "Payment Method" column in the dataset ensures that every entry in this column adheres to the set of predefined valid payment methods. Additionally, it can also be used as pre checker of the fact that all the rows of the column "Payment Method" has been filled with correct values and has not been left empty.

### **Cell 2: Parameters for checker**

In [24]:
# params for presence error
expected_payment_methods = ['Cash', 'Credit Card', 'Debit Card', 'Digital Wallet']


### **Cell 3: Checker code**

In [25]:
#fund to check payment method has been selected correctly
def check_payment_method(row, expected_methods):
  payment_method = str(row['Payment Method']).strip()

  if not payment_method:
    return f"Empty payment method: {payment_method}"

  if payment_method not in expected_methods:
    return f"Invalid payment method: {payment_method}"
  return None

#apply the func to the dataset for the payment methid presence check
df['Payment Method Check'] = df.apply(lambda row: check_payment_method(row, expected_payment_methods), axis=1)

#Filter the invalid payment method
invalid_payment_methods = df[df['Payment Method Check'].notna()]

 # Display findings
print("Invalid Payment Methods:")
print(invalid_payment_methods[['Transaction ID', 'Payment Method', 'Payment Method Check']])
if len(invalid_payment_methods) > 0:
  print(f"There are {len(invalid_payment_methods)} invalid payment methods.")
else:
  print("No invalid payment methods found.")



Invalid Payment Methods:
     Transaction ID Payment Method             Payment Method Check
3       TXN_7034554        UNKNOWN  Invalid payment method: UNKNOWN
6       TXN_4433211          ERROR    Invalid payment method: ERROR
8       TXN_4717867            NaN      Invalid payment method: nan
9       TXN_2064365            NaN      Invalid payment method: nan
13      TXN_9437049            NaN      Invalid payment method: nan
...             ...            ...                              ...
9985    TXN_3297457            NaN      Invalid payment method: nan
9988    TXN_9594133          ERROR    Invalid payment method: ERROR
9992    TXN_2739140        UNKNOWN  Invalid payment method: UNKNOWN
9994    TXN_7851634            NaN      Invalid payment method: nan
9995    TXN_7672686            NaN      Invalid payment method: nan

[3178 rows x 3 columns]
There are 3178 invalid payment methods.


### **Cell 4: Report of findings**
The presence error check for payment methods has revealed a significant number of inconsistencies within the dataset. A total of 3,178 transactions have been flagged with invalid payment methods. These discrepancies are characterized by unrecognized entries labeled as 'UNKNOWN', 'ERROR', or simply missing data (displayed as 'NaN' or blank entries). This level of inconsistency not only compromises the integrity of the transaction data but also poses challenges for accurate financial reporting and analysis. The output indicates the nessicity of stricter data validation measures such as presence error and correcting the existing discrepancies for ensuring accurate and meaningful transactions

## Error no:7 (Length errors)

### **Cell 1: Test Name and description**
  * Error name: **Length Error**
  * Error description: Length check error ensures that an appropriate number of characters are entered into the field. It verifies that neither too long nor too short values is eneterd into the field that can impact the consistencies of the dataset.
  
**Note:** For the purpose of this assignment, we are using Transaction ID to visualize the Length Check validation since no other column in the dataset seemed appropriate for this check. Usually, telephone num, and password fields are used for length check validation; which were not present in the dataset. Therefore, we are using Transaction ID. Also, in most cases Transaction ID are automatically generated by the ssytem.

### **Cell 2: Parameters for checker**

In [26]:
#param for length check
expected_transaction_id_length = 11

### **Cell 3: Checker code**

In [28]:
#func to check transaction ID length
def check_transaction_id_length(row, expected_length):
  transaction_id = str(row['Transaction ID']).strip()
  actual_transaction_id_length = len(transaction_id)
  if actual_transaction_id_length != expected_length:
    return f"Expected length: {expected_length}, Found length: {actual_transaction_id_length}"
  if not transaction_id:
    return f"Empty transaction ID: {transaction_id}"
  return None

#apply the func to each row of the dataset for column Transaction ID
df['Transaction ID Length Check'] = df.apply(lambda row: check_transaction_id_length(row, expected_transaction_id_length), axis=1)

#filter out the invalid length
invalid_transaction_ids = df[df['Transaction ID Length Check'].notna()]

# Display findings
print("Invalid Transaction IDs:")
print(invalid_transaction_ids[['Transaction ID', 'Transaction ID Length Check']])
if len(invalid_transaction_ids) > 0:
  print(f"There are {len(invalid_transaction_ids)} invalid transaction IDs.")
else:
  print("No INVALID transaction IDs found.")

Invalid Transaction IDs:
Empty DataFrame
Columns: [Transaction ID, Transaction ID Length Check]
Index: []
No INVALID transaction IDs found.


### **Cell 4: Report of findings**
According to the generated output of the length validation checker, the transaction IDs in the dataset did not identify any discrepancies. Each transaction ID accurately meets the expected length criterion, which maintains data uniformity and integrity for this key identifier. This outcome confirms that the transaction ID data entry processes are currently effective, ensuring consistent and reliable data across the dataset. The absence of length errors suggests that the controls in place for data input are robust, but continuous monitoring remains essential to maintain this standard.

If there there were any rows under transaction ID with wrong length,  the length checker could have identified that and showed them as an output.


## Error no: 8 (Look-up errors)

### **Cell 1: Test Name and description**
  * Error name: **Look-up Error**
  * Error description: This error occurs when entries in the 'Location' column do not match the predefined list of accepted location names or are left blank. Invalid or missing location data can lead to inconsistencies in reporting and analysis, affect logistical planning, and potentially disrupt operational efficiencies. Employing a lookup checker as a preliminary check ensures that all entries are either accurate or flagged for correction, thereby preventing any rows from being left empty or incorrectly filled with labels such as 'unknown' or 'error.'r


### **Cell 2: Parameters for checker**

In [29]:
# params for look up error
expected_locations = ['Takeaway','In-store']

### **Cell 3: Checker coder**

In [30]:
#func to check look-up errors
def check_location(row, expected_locations):
  location = str(row['Location']).strip()
  if not location:
    return f"Empty location: {location}"
  if location not in expected_locations:
    return f"Invalid location: {location}"
  return None

#apply the func to check the look-up errors
df['Location Look-up Check'] = df.apply(lambda row: check_location(row, expected_locations), axis=1)

#filter out invalid entries from the dataframe
invalid_locations = df[df['Location Look-up Check'].notna()]

# Display findings
print("Invalid Locations:")
print(invalid_locations[['Transaction ID', 'Location', 'Location Look-up Check']])
if len(invalid_locations) > 0:
  print(f"There are {len(invalid_locations)} invalid locations.")
else:
  print("No invalid locations found.")



Invalid Locations:
     Transaction ID Location     Location Look-up Check
3       TXN_7034554  UNKNOWN  Invalid location: UNKNOWN
5       TXN_2602893      NaN      Invalid location: nan
7       TXN_6699534  UNKNOWN  Invalid location: UNKNOWN
16      TXN_3765707      NaN      Invalid location: nan
18      TXN_8876618    ERROR    Invalid location: ERROR
...             ...      ...                        ...
9994    TXN_7851634      NaN      Invalid location: nan
9995    TXN_7672686  UNKNOWN  Invalid location: UNKNOWN
9996    TXN_9659401      NaN      Invalid location: nan
9997    TXN_5255387      NaN      Invalid location: nan
9998    TXN_7695629      NaN      Invalid location: nan

[3961 rows x 3 columns]
There are 3961 invalid locations.


### **Cell 4: report of findings**
From the generated output above we can see that, the dataset's location data reveals a significant number of discrepancies, with 3,961 transactions flagged for having invalid locations. The presence of numerous 'UNKNOWN', 'ERROR', and missing ('NaN') location values necessitates a comprehensive audit and correction initiative.These issues are largely characterized by unrecognized location entries and empty fields. This situation highlights potential flaws in the data entry processes or mismatches in location data synchronization. By using a look-up checker like above, it can be ensured that no invalid data is placed in the dataset.

## Error no: 9 (Exact duplicate errors)  

**Note: "Movies.csv" data file has been used for this test checking**



### **Cell 1: Test Name and description**
  * [**Note**: "Movies.csv" data file has been used for this since the other  dataset do not have exact duplicates.]
  * Error name: **Exact duplicate Error**
  * Error description: The exact duplicate checker identifies rows in the dataset that are exact duplicates across all selected fields. Exact duplicates can lead to data redundancy, potentially skewing data analysis and reporting. Removing these duplicates is crucial for maintaining the accuracy and reliability of our dataset.


### **Cell 2: Parameters for checker**

In [32]:
#copy the actual dataset so that it does not changes
data_movies_copy = data_movies.copy()
#cols to drop which are not necessary for this exact duplicate checker
col_to_drop = ['RATING', 'VOTES']
#drop the specified cols
data_movies_copy.drop(columns=col_to_drop, inplace=True)

### **Cell 3: Checker code**

In [33]:
# checking for duplicate rows
duplicates_movie_exact = data_movies_copy.duplicated(keep = False)

# Extracting duplicate rows to review them
data_movies_copy[duplicates_movie_exact]

# Displaying the duplicates.
print("There are ",len(data_movies_copy[duplicates_movie_exact]),"duplicate entries in the dataset")
print("The first 20 exact duplicate rows are as follows:")

# Extracting duplicate rows to review them
data_movies_copy[duplicates_movie_exact].head(20)


There are  662 duplicate entries in the dataset
The first 20 exact duplicate rows are as follows:


,MOVIES,YEAR,GENRE,ONE-LINE,STARS,RunTime,Gross
6809,Brigada Costa del Sol,(2019),"\nAction, Adventure, Drama",\nAdd a Plot\n,\n Director:\nNorberto López Amado\n| \n ...,NaN,NaN
6830,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",NaN,NaN
6831,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",NaN,NaN
6832,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",NaN,NaN
6833,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",NaN,NaN
6834,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",NaN,NaN
6835,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",NaN,NaN
6836,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",NaN,NaN
6837,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",NaN,NaN
6838,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",NaN,NaN


### **Cell 4: Report of findings**
The duplicate check has identified several exact duplicate rows within the dataset. These duplicates, if left unchecked, could distort analyses such as trend estimation and statistical calculations. The findings have prompted a review of data entry processes to prevent such occurrences in the future. Depending on the review outcomes, duplicates may be removed to ensure the dataset's integrity, or kept if they serve a specific purpose.


## Error no: 10 (Near duplicate errors)  
**Note: "Movies.csv" data file has been used for this test**

### **Cell 1: Test Name and description**
  * [**Note**: "Movies.csv" data file has been used for this since the other  dataset do not have exact duplicates.]
  * Error name: **Near duplicate Error**
  * Error description: The near duplicate error dentifies rows in the dataset that are nearly identical across most fields except for two columns such as 'RATING' and 'VOTES'. Near duplicates can occur due to minor variations in data entry, partial updates to records, or errors in data import.

### **Cell 2: Parameters for checker**

In [34]:
# checking for duplicate rows
duplicates_movie_near = data_movies.duplicated(keep = False)

### **Cell 3: Checker code**

In [35]:
# Extracting duplicate rows to review them
data_movies[duplicates_movie_near]

# Displaying the duplicates
print("There are approximately ",len(data_movies[duplicates_movie_near]),"near duplicate entries in the dataset")
print("The first 20 near duplicate rows are as follows:")
# Extracting duplicate rows to review them
data_movies[duplicates_movie_near].head(10)

There are approximately  548 near duplicate entries in the dataset
The first 20 near duplicate rows are as follows:


,MOVIES,YEAR,GENRE,RATING,ONE-LINE,STARS,VOTES,RunTime,Gross
6830,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",9.0,\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",6,NaN,NaN
6832,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",NaN,\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",NaN,NaN,NaN
6833,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",NaN,\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",NaN,NaN,NaN
6835,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",9.0,\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",6,NaN,NaN
6836,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",9.0,\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",6,NaN,NaN
6837,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",NaN,\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",NaN,NaN,NaN
6838,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",NaN,\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",NaN,NaN,NaN
6839,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",NaN,\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",NaN,NaN,NaN
6840,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",NaN,\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",NaN,NaN,NaN
6841,Mighty Little Bheem,(2019– ),"\nAnimation, Short, Adventure",NaN,\nAdd a Plot\n,"\n Directors:\nRajiv Chilaka, \nKrishna Moh...",NaN,NaN,NaN


### **Cell 4: Report of findings**
The analysis identified several near duplicate entries in the dataset. As we can see in the above generated output that most of the fields are almost similar except for two columns ('RATINGS', 'VOTES'). Identidying near duplicate is not an easy task. It depends on the use case of the analysis, data domain on which analysis is being conducted, and overall variation in the dataset. Near duplicates orginates from  discrepancies in data collection or update processes. Addressing these near dulicates will not only streamline data aggregation but also enhance data quality, supporting more reliable analyses and decisions.


## Conclusion ##
This report contains the suggested types of clean data checks. Different columns of dataset has been used to perform the different types of data cleaning checkers. Exact duplicate and Near duplicate data checks were performed using the Movies dataset as the first dataset was not appropriate for the required clean data checks. A further improvement can be done by adding more user interactive checkers for various features rather than statically setting up the checkers.

## References


*   https://www.geeksforgeeks.org/how-to-convert-strings-to-floats-in-pandas-dataframe/
*   https://www.geeksforgeeks.org/python-pandas-to_numeric-method/


*   https://www.influxdata.com/blog/python-date-comparison-comprehensive-tutorial/#:~:text=You%20can%20use%20the%20strptime%20function%20for%20this%20purpose.&text=In%20this%20example%2C%20the%20format%20string%20%E2%80%9C%25Y%2D%25m,the%20date%20and%20format%20strings.
*   https://medium.com/towards-data-science/practical-guide-to-data-analysis-and-preprocessing-080815548173


*   https://sparkbyexamples.com/pandas/pandas-get-day-month-and-year-from-datetime/







